azurerm/resource_arm_network_watcher_flow_log.go

```diff
@@ -0,0 +1,350 @@
+	}
+
+	resp, err := client.Get(ctx, resourceGroupName, networkWatcherName)
+	if err != nil {
+		return fmt.Errorf("Cannot read Network Watcher %q (Resource Group %q) ID: %+v", networkWatcherName, resourceGroupName, err)
+	}
+	d.SetId(*resp.ID)
+
+	return resourceArmNetworkWatcherFlowLogRead(d, meta)
+}
+
+func resourceArmNetworkWatcherFlowLogRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).watcherClient
+	ctx := meta.(*ArmClient).StopContext
+
+	id, err := parseAzureResourceID(d.Id())
+	if err != nil {
+		return err
+	}
+	resourceGroupName := id.ResourceGroup
+	networkWatcherName := id.Path["networkWatchers"]
+
+	// Get current flow log status
+	networkSecurityGroupID := d.Get("network_security_group_id").(string)
+	statusParameters := network.FlowLogStatusParameters{
+		TargetResourceID: &networkSecurityGroupID,
+	}
+	future, err := client.GetFlowLogStatus(ctx, resourceGroupName, networkWatcherName, statusParameters)
+	if err != nil {
+		if strings.Contains(err.Error(), "StatusCode=0 -- Original Error: Code=\"NotFound\"") {
+			// One of storage account, NSG, or flow log is missing
+			log.Printf("[INFO] Error getting Flow Log Configuration %q for target %q - removing from state", d.Id(), networkSecurityGroupID)
+			d.SetId("")
+			return nil
+		}
+
+		return fmt.Errorf("Error retrieving Flow Log Configuration for target %q (Network Watcher %q / Resource Group %q): %+v", networkSecurityGroupID, networkWatcherName, resourceGroupName, err)
+	}
+
+	if err = future.WaitForCompletionRef(ctx, client.Client); err != nil {
+		return fmt.Errorf("Error waiting for retrieval of Flow Log Configuration for target %q (Network Watcher %q / Resource Group %q): %+v", networkSecurityGroupID, networkWatcherName, resourceGroupName, err)
+	}
+
+	fli, err := future.Result(client)
+	if err != nil {
+		return fmt.Errorf("Error retrieving Flow Log Configuration for target %q (Network Watcher %q / Resource Group %q): %+v", networkSecurityGroupID, networkWatcherName, resourceGroupName, err)
+	}
+
+	d.Set("network_watcher_name", networkWatcherName)
+	d.Set("resource_group_name", resourceGroupName)
+
+	d.Set("network_security_group_id", fli.TargetResourceID)
+	d.Set("enabled", fli.FlowLogProperties.Enabled)
+
+	// Azure API returns "" when flow log is disabled
+	// Don't overwrite to prevent storage account ID diff when that is the case
+	if *fli.FlowLogProperties.StorageID != "" {
+		d.Set("storage_account_id", fli.FlowLogProperties.StorageID)
+	}
+
+	if err := d.Set("retention_policy", flattenAzureRmNetworkWatcherFlowLogRetentionPolicy(fli.FlowLogProperties.RetentionPolicy)); err != nil {
+		return fmt.Errorf("Error setting `retention_policy`: %+v", err)
+	}
+
+	if err := d.Set("traffic_analytics", flattenAzureRmNetworkWatcherFlowLogTrafficAnalytics(fli.FlowAnalyticsConfiguration)); err != nil {
+		return fmt.Errorf("Error setting `traffic_analytics`: %+v", err)
+	}
```

Comment:
since the `properties` block may not necessarily be returned (due to an invalid API response) - could we wrap this in an if statement, e.g.:

```
if props := fli.FlowKLogProperties; props != nil {
  d.Set("network_security_group_id", props.TargetResourceID)
  d.Set("enabled", props.Enabled)

  if props.StorageID != nil && *props.StorageID != "" {
    d.Set("storage_account_id", props.StorageID)
  }

  if err := d.Set("retention_policy", flattenAzureRmNetworkWatcherFlowLogRetentionPolicy(props.RetentionPolicy)); err != nil {
    return fmt.Errorf("Error setting `retention_policy`: %+v", err)
  }

  if err := d.Set("traffic_analytics", flattenAzureRmNetworkWatcherFlowLogTrafficAnalytics(props.FlowAnalyticsConfiguration)); err != nil {
    return fmt.Errorf("Error setting `traffic_analytics`: %+v", err)
  }
}
```